text/2094-nll.md

```diff
@@ -1592,7 +1592,49 @@ our example, this would be the case if the type of `list` had been
 reported by the next portion of the borrowck, described in the next
 section.
 
-### Borrow checker phase 2: reporting errors
+### Borrow Checker phase 2: computing initialization
+
+In addition to propagating borrows, the borrow-checker makes sure that uninitialized values can never be seen.
+
+To ensure that, we also compute whether each lvalue is potentionally and/or definitely initialized at each point in the CFG.
+
+We compute that using another standard dataflow computation. Of course, there might be an infinite amount of lvalues (e.g. consider a linked list using `Box`), but the computation can be done in finite time because lvalues form a tree, and the transfer function always affects an **lvalue subtree** at once (where an lvalue `cv` is in the subtree generated by the lvalue `lv` if `lv` is a **prefix** of `cv`).
+
+The transfer function is as follows:
+
+- at function entry, the function arguments are definitely initialized, and all other locals are definitely uninitialized
+- after a consuming use of the lvalue `lv`, if the type of `lv` does not implement `Copy`, all lvalues in the lvalue subtree of `lv` are definitely uninitialized
+- after a MIR `drop(lv)` or `storagedead lv` statement, all lvalues in the lvalue subtree of `lv` are definitely uninitialized
+- after an `lv = <rvalue>` statement or an `lv <- <rvalue>` (drop+replace) statement, all lvalues in the lvalue subtree of `lv` are definitely initialized
+- after a non-panic return from an `lv = call <fn>(<args>)` statement, all lvalues in the lvalue subtree of `lv` are definitely initialized
+
+NOTE: it is possible for all the fields of a struct or tuple to be initialized without the struct/tuple itself being initialized, for example:
+```Rust
+let x: (u32, u32);
+x.0 = 4;
+x.1 = 5;
```

Comment:
In the current borrow checker, [we do not accept this program](https://play.rust-lang.org/?gist=9f5087a9c82fbf836013615c5d9f2429&version=stable). Are there corner cases that I am overlooking here? 

Presuming there are not, I think that the current borrow checker basically enforces a rule like "Assigning to an lvalue LV requires that all prefixes of LV are definitely initialized", right?

Maybe we should just adopt this rule for now?

Reply:
This is indeed a difference from the current borrow checker. I see no problem with this rule.